File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def solution(mats, park):
    answer = -1
    n = len(park)
    m = len(park[0])
    board = [[0]*m for _ in range(n)]
    
    for i in range(len(park)):
        for j in range(len(park[0])):
            if park[i][j] =="-1":
                board[i][j] = 1
            else:
                board[i][j] =0
    
    dp = [[0]*m for _ in range(n)]
    max_size = 0
    for i in range(len(park)):
        for j in range(len(park[0])):
            if board[i][j] == 1:
                if i == 0 or j ==0:
                    dp[i][j] = 1
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1           
                max_size = max(max_size, dp[i][j])
    
    mats.sort(reverse = True)
    for i in mats:
        if i <= max_size:
            return i
    return answer

File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.java
import java.util.*;


class Solution {
    public int solution(int[] mats, String[][] park) {
        int ms = 1;
        int n =park.length;
        int m = park[0].length;
        int [][] board = new int[n][m];
        int [][] dp = new int[n][m];
        for(int i = 0; i < n; i++){
            for(int j = 0; j<m;j++){
                if(park[i][j].equals("-1")){
                    board[i][j] = 1;
                }
            }
        }
        
        for(int i = 0; i< n;i++){
            for(int j = 0; j< m;j++){
                if(board[i][j] == 1){
                    if(i==0||j==0){
                        dp[i][j]= 1;
                    }else{
                        dp[i][j] = Math.min(dp[i-1][j-1], Math.min(dp[i-1][j],dp[i][j-1]))+1;
                    }
                    ms = Math.max(dp[i][j],ms);
                }
            }
        }
        int bestSize = -1;
        for (int size : mats) {
            if (size <= ms && size > bestSize) {
                bestSize = size;
            }
        }
        return bestSize;
    }
}